File: IT/1/Eccezioni/Esercizi/C++/esercizio13.cpp
/*
file esercizio13.cpp
compilare con: g++ -std=c++11 esercizio13.cpp
*/

#include <iostream>
#include <cmath>
#include <stdexcept>

using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::string;
using std::invalid_argument;
using std::domain_error;

int main()
{
	double l = -1;

	while (l < 0) {
		cout << "Scrivi un numero: ";
		string s;
		getline(cin, s);

		try
		{
			int n = stoi(s);

			if (n <= 0) {
				throw domain_error("Errore di dominio matematico: logaritmo non definito per n <= 0");
			}

			l = log(n);

			cout << "Logaritmo di " << n << " = " << l << endl;
		}
		catch (invalid_argument &e) {
			cerr << e.what() << endl;
		}
		catch (domain_error &e) {
			cerr << e.what() << endl;
		}
	}

	cout << "Fine del programma." << endl;
}

File: IT/1/Eccezioni/Esercizi/Java/Esercizio13.java
/* file Esercizio13.java */

public class Esercizio13 {
	public static void main(String [] args) {
		double l = -1;

		while (l < 0) {
			String s = System.console().readLine("Scrivi un numero: ");

			try {
				int n = Integer.parseInt(s);

				if (n <= 0) {
					throw new IllegalArgumentException("Errore di dominio matematico: logaritmo non definito per n <= 0");
				}

				l = Math.log(n);
				System.out.println("Logaritmo di " + n + " = " + l);
			}
			catch (NumberFormatException e) {
				System.out.println(e.getMessage());
			}
			catch (IllegalArgumentException e) {
				System.out.println(e.getMessage());
			}
		}

		System.out.println("Fine del programma.");
	}
}

